modules/apps/healthcheck/healthcheck-api/src/main/java/com/liferay/healthcheck/HealthcheckItem.java

```diff
@@ -0,0 +1,85 @@
+/**
+ * SPDX-FileCopyrightText: (c) 2024 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+package com.liferay.healthcheck;
+
+import com.liferay.petra.string.StringUtil;
+import com.liferay.portal.kernel.util.ArrayUtil;
+
+/**
+ * @author Olaf Kock
+ */
+public class HealthcheckItem {
+
+	public HealthcheckItem(
+		boolean resolved, String link, String messageKey,
+		Object... messageParameters) {
+
+		_resolved = resolved;
+		_link = link;
+		_messageKey = messageKey;
+		_messageParameters = messageParameters;
+	}
+
+	/**
+	 * A link(URL) that can contain further information on the tested condition
+	 *
+	 * @return a link URL
+	 */
+	public String getLink() {
+		return _link;
+	}
+
+	/**
+	 * A message key that describes what this healthcheck checks for
+	 *
+	 * @return human readable message
+	 */
+	public String getMessageKey() {
+		return _messageKey;
+	}
+
+	/**
+	 * Parameters that are used to look up the human readable message,
+	 * with the messageKey.
+	 * @return an array of parameters to be combined with the localized message
+	 */
+	public Object[] getMessageParameters() {
+		return _messageParameters;
+	}
+
+	/**
+	 * A machine readable key that can be used to refer to a particular healthcheck
+	 * or its result. This was introduced to be able to ignore certain healthchecks,
+	 * in case their test does not apply to a certain environment (example:
+	 * Elasticsearch Sidecar is ok in local demo systems). Default content: The
+	 * healthcheck's fully qualified classname, optionally extended by extra
+	 * information (each healthcheck might execute several checks) and the
+	 *
+	 * @return the machine readable encoding for this healthcheck
+	 */
+	public String getSourceKey() {
```

Comment:
@olafk this can not be right, you use the sourcekey as the ignore key. but source key is including moving piece of info from messageparameters. From run the run if the messageparameters change, then this HealthcheckItem won't match previously ignored result.

For example, here https://github.com/shuyangzhou/liferay-portal/pull/11452/commits/c1763a2f52169a8f337bbb1742b1316c35ea4686#diff-e6b362b99e00775d436cfdee7d2eaca863aff2299f9bbcf44001cce99089d6eeR42 , if I ignore it with current value,  and restart the server with a different settings, then it won't match the ignored entry anymore. So think about, if the new setting is still not following the suggestion, then user will be warned again, even they previously ignored it. And if they ignore the new warn again, we will end up 2 ignored entries for the same item.

I don't really think you have a clean way to ignore items based on parameters. And if you don't include parameters, isn't that just going to always tie to the healthcheck itself? I think it is better just to control ignore at the healthcheck level, rather than item level.

Reply:
There are some particular healthchecks that rely on arbitrary additional data for being ignorable _in a helpful way_: E.g. the (4th commit) Client-Extension and Data-Provider hostname checks also include the current system's virtualhostname: This way, they will also trigger when a PRD backup is restored on UAT and the system's virtual host name changes. 

_This_ kind of triggering is what makes the healthchecks _really_ valuable IMHO, and they are what got me started implementing them: In SalesEngineering, we don't go from PRD to UAT, but from a template demo to an instantiated demo with different host names.

I'd be happy to demonstrate this in a call (I already have the virtual-host-name validator implemented. As you mentioned you don't want to instantiate another default filter, I have made it optional: Once activated, this kind of environment-dependent validation can help catching subtle bugs, like PRD referencing UAT environments, or vice versa. This is something that can't be done through generic whitelists.

Reply:
the bottom line, for anything that could be ignored, it must be identifiable in a stable way, meaning from run to run (different jvms with different settings) it must be seen as the same item. Otherwise, we will run into orphan data and duplicated ignored items.

If you can not just do the ignore at check level, then not all items should support ignore, only those stably identifiable ones should.

Reply:
There are 2 possible ways for this.
1) Just drop the ignore function, at least for now. To make some progress on this pull.
2) Don't run the checks on page rendering, change it to a button click, and save the entire check results into preferences. When you ignore items, just update the saved results, when render the page just display the previous results. User can always click the button again to rerun, which will throw away the previously saved results, including the ignored items.

Reply:
Thank you for the feedback.
I'll be looking into it, but as you could see on the lag: I've been busy with other tasks, and couldn't come back to it.
I'm hesitant to take out the "ignore" function, as it (together with system-dependent keys) provides quite a huge amount of the added value, e.,g. when a PRD system is restored in UAT.
If this PR is open for too long and blocks your queue, feel free to reject it, and I'll PR again once I came to a conclusion to either remove or enhance the current feature set, and/or come with a good use case for it in written form.

Reply:
Hi Shuyang,
I hope you remember a bit from this PR and our discussion still...
Finally, I've looked at this PR again and tried to digest all the suggestions.

With the way I'm using healthchecks already in my SE work (as a custom plugin), I really rely on the "ignore" features and the quick way to clean up the list.

The way the current "ignore" function is implemented is **100% within the UI that we agreed _not_ to include in master** - rather the UI-commit is a throwaway commit that enables _you_ to look at the healthchecks and see what the result/vision is. This has a few ramifications with regards to the your comments and the "ignore" feature:

* There still is the option to implement blacklisting on a class-level - if I really want to ignore a healthcheck implementation no matter what the output is. 
* Asynchronous calling - e.g. detached from rendering the portlet - is still possible to implement with a different frontend. Accepting the PR without the UI as agreed would result in code in core that is never called anyway - neither synchronously nor asynchronously.
* The actual "ignore" implementation _could_ be implemented to self-clean (automatically or manually), e.g. to throw away orphan keys when they don't apply to the currently found issues. Either way, the current "ignore" function is located in the throwaway-UI that's not supposed to make it into core anyway.
* On top, the amount of orphan data is quite low: It's a small number of strings: If it was a truly remarkable amount of data, that would mean that a truly remarkable number of healthchecks have failed before - and in that case, it would be easier to ignore them by not looking at them anyway.

Some more practical description: In my Sales-Engineering day-job, I constantly restore demo systems prepared by someone else, under a different name. Specifically with regards to Client Extensions and DataProviders, I rely on healthchecks as a custom plugin: They often refer to the original host name, not to the host name that I use for my upcoming demo. But sometimes they reference a release-file from github, which is fine to ignore: This is a prime case for ignoring results based on the actual parameter.

I hope this long list of reasons makes it possible to look at the (unchanged) commits again. Let me know if it's worth to rebase and I'll happily do so. 

As of the code/features, I'm sure I'd like to see it to be continued the way it currently is. If you can come up with clearer naming (e.g. for the "sourceKey", but not limited to it), I'd definitely be happy to change that.